pkg/services/ngalert/notifier/channels/slack.go

```diff
@@ -35,132 +35,94 @@ type SlackNotifier struct {
+	return ch, nil
 }
 
-func NewSlackConfig(factoryConfig FactoryConfig) (*SlackConfig, error) {
+func buildSlackNotifier(factoryConfig FactoryConfig) (*SlackNotifier, error) {
 	channelConfig := factoryConfig.Config
 	decryptFunc := factoryConfig.DecryptFunc
-	endpointURL := channelConfig.Settings.Get("endpointUrl").MustString(SlackAPIEndpoint)
-	slackURL := decryptFunc(context.Background(), channelConfig.SecureSettings, "url", channelConfig.Settings.Get("url").MustString())
+	var settings slackSettings
+	err := factoryConfig.Config.unmarshalSettings(&settings)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.EndpointURL == "" {
+		settings.EndpointURL = SlackAPIEndpoint
+	}
+	slackURL := decryptFunc(context.Background(), channelConfig.SecureSettings, "url", settings.URL)
 	if slackURL == "" {
-		slackURL = endpointURL
+		slackURL = settings.EndpointURL
 	}
 	apiURL, err := url.Parse(slackURL)
 	if err != nil {
 		return nil, fmt.Errorf("invalid URL %q", slackURL)
 	}
-	recipient := strings.TrimSpace(channelConfig.Settings.Get("recipient").MustString())
-	if recipient == "" && apiURL.String() == SlackAPIEndpoint {
+	settings.URL = apiURL.String()
+
+	settings.Recipient = strings.TrimSpace(settings.Recipient)
+	if settings.Recipient == "" && settings.URL == SlackAPIEndpoint {
 		return nil, errors.New("recipient must be specified when using the Slack chat API")
 	}
-	mentionChannel := channelConfig.Settings.Get("mentionChannel").MustString()
-	if mentionChannel != "" && mentionChannel != "here" && mentionChannel != "channel" {
-		return nil, fmt.Errorf("invalid value for mentionChannel: %q", mentionChannel)
+	if settings.MentionChannel != "" && settings.MentionChannel != "here" && settings.MentionChannel != "channel" {
+		return nil, fmt.Errorf("invalid value for mentionChannel: %q", settings.MentionChannel)
 	}
-	token := decryptFunc(context.Background(), channelConfig.SecureSettings, "token", channelConfig.Settings.Get("token").MustString())
-	if token == "" && apiURL.String() == SlackAPIEndpoint {
+	settings.Token = decryptFunc(context.Background(), channelConfig.SecureSettings, "token", settings.Token)
+	if settings.Token == "" && settings.URL == SlackAPIEndpoint {
 		return nil, errors.New("token must be specified when using the Slack chat API")
 	}
-	mentionUsersStr := channelConfig.Settings.Get("mentionUsers").MustString()
-	mentionUsers := []string{}
-	for _, u := range strings.Split(mentionUsersStr, ",") {
-		u = strings.TrimSpace(u)
-		if u != "" {
-			mentionUsers = append(mentionUsers, u)
-		}
+	if settings.Username == "" {
+		settings.Username = "Grafana"
 	}
-	mentionGroupsStr := channelConfig.Settings.Get("mentionGroups").MustString()
-	mentionGroups := []string{}
-	for _, g := range strings.Split(mentionGroupsStr, ",") {
-		g = strings.TrimSpace(g)
-		if g != "" {
-			mentionGroups = append(mentionGroups, g)
-		}
+	if settings.Text == "" {
+		settings.Text = `{{ template "default.message" . }}`
```

Comment:
let's declare constant in `default_template.go` 
```
const DefaultMessageEmbed = `{{ template "default.message" .}}`
```

Reply:
I'll put up a separate PR for this one, I think that'll be quicker than extracting them one-by-one

Reply:
Done in a separate PR: https://github.com/grafana/grafana/pull/55424
Now we don't have to remember to do this for each notifier we touch